### PR TITLE
Refactoring : Unify group type check in ResourceMapper extractField

### LINKED ISSUE #797: Cleanup : Unify group type check in ResourceMapper extractField method.
**Describe the Issue**
Unify group type check which currently present twice in extractField method.

**Review Comment**
` i think this needs a bit of cleanup. line 233 and line 260 both check if the item is a group. can this logic be unified rather than scatted around in different places? `

https://github.com/google/android-fhir/pull/676/files#r713841089 

**Would you like to work on the issue?**

### PR DESCRIPTION
**Pull request by santosh-pingle**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #797

**Description**
Refactoring : Unify group type check in ResourceMapper extractField

**Alternative(s) considered**
no

**Type**
Choose one: Code health

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 1]
```diff
@@ -244,59 +244,65 @@ object ResourceMapper {
...
+  ) {
+    if (questionnaireItem.itemContextNameToExpressionMap.values.isEmpty()) {
+      return
+    }
+    val resource = questionnaireItem.createBase()
+    if (resource !is Resource) {
+      return
+    }
+    resource.apply { extractFields(bundle, questionnaireItem.item, questionnaireResponseItem.item) }
+    bundle.apply { addEntry().apply { this.resource = resource } }
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:305]: no need to use `apply` here

could just be `resource.extractFields` and `bundle.addEntry`...

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 2]
```diff
@@ -545,6 +551,11 @@ private fun Questionnaire.QuestionnaireItemComponent.getFieldOrNull(
   }
 }
 
+private fun Questionnaire.QuestionnaireItemComponent.createBase(): Base =
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:554]: add documentation

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 3]
```diff
@@ -244,59 +244,65 @@ object ResourceMapper {
...
+    if (!definitionField.nonParameterizedType.isEnum) {
+      // this is a low level type e.g. StringType
+      updateField(definitionField, questionnaireResponseItem.answer)
+      return
     }
+    // this is a high level type e.g. AdministrativeGender
+    updateFieldWithEnum(definitionField, questionnaireResponseItem.answer.first().value)
+  }
+
+  private suspend fun extractResource(
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:292]: add documentation

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 4]
```diff
@@ -244,59 +244,65 @@ object ResourceMapper {
...
-        updateFieldWithEnum(definitionField, questionnaireResponseItem.answer.first().value)
-      }
+    if (questionnaireResponseItem.answer.isEmpty()) return
+    if (!definitionField.nonParameterizedType.isEnum) {
+      // this is a low level type e.g. StringType
+      updateField(definitionField, questionnaireResponseItem.answer)
+      return
     }
+    // this is a high level type e.g. AdministrativeGender
+    updateFieldWithEnum(definitionField, questionnaireResponseItem.answer.first().value)
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt]: Not sure about low level and high level types which were present as inline comment in existing code.
Not sure about FHIR standard which has mention of it.

It seems comment `// this is a high level type e.g. AdministrativeGender` try to say about these are value set and represented by enum. [valuesets](https://www.hl7.org/fhir/terminologies-valuesets.html)
Is it okay to mention comment below?
_Value set defined as part of the FHIR specification, which include codes, https://www.hl7.org/fhir/terminologies-valuesets.html_

And inline comment `// this is a low level type e.g. StringType` may be suggests primitive type in FHIR.
[link] (https://www.hl7.org/fhir/datatypes.html#primitive)
Is it okay to mention comment below?
_Primitive types are those with only a value, and no additional elements as children, https://www.hl7.org/fhir/datatypes.html#primitive_

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 5]
```diff
@@ -244,60 +244,71 @@ object ResourceMapper {
...
-      } else {
-        val value: Base =
-          (definitionField.nonParameterizedType.newInstance() as Base).apply {
-            extractFields(bundle, questionnaireItem.item, questionnaireResponseItem.item)
-          }
-        updateField(definitionField, value)
-      }
+    if (questionnaireResponseItem.answer.isEmpty()) return
+    if (definitionField.nonParameterizedType.isEnum) {
+      // this is a high level type e.g. AdministrativeGender
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:284]: just remove this comment

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 6]
```diff
@@ -244,60 +244,71 @@ object ResourceMapper {
...
     } else {
-      if (questionnaireResponseItem.answer.isEmpty()) return
-      if (!definitionField.nonParameterizedType.isEnum) {
-        // this is a low level type e.g. StringType
-        updateField(definitionField, questionnaireResponseItem.answer)
-      } else {
-        // this is a high level type e.g. AdministrativeGender
-        updateFieldWithEnum(definitionField, questionnaireResponseItem.answer.first().value)
-      }
+      // this is a low level type e.g. StringType
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:287]: and remove this comment too

**Review by jingtang10 — APPROVED** [review]

**Pull request merged**